### PR TITLE
bpo-41835: Optimize dict() call with keyword arguments

### PR DESCRIPTION
**Pull request by methane**
https://bugs.python.org/issue41835

### CODE REVIEW

File: Objects/dictobject.c [hunk 1]
```diff
@@ -1650,6 +1650,69 @@ PyDict_SetItem(PyObject *op, PyObject *key, PyObject *value)
...
+
+    for (Py_ssize_t i = 0; i < size; i++) {
+        PyObject *key = PyTuple_GET_ITEM(keys, i);
+        Py_INCREF(key);
+        if (!PyUnicode_CheckExact(key)) {
+            all_unicode = 0;
+        }
+
+        Py_hash_t hash;
+        if (!PyUnicode_CheckExact(key) ||
```
Comment (Marco-Sulla) [inline Objects/dictobject.c:1683]: PyUnicode_CheckExact is done above, you can store it in a var.

File: Objects/dictobject.c [hunk 2]
```diff
@@ -1650,6 +1650,69 @@ PyDict_SetItem(PyObject *op, PyObject *key, PyObject *value)
...
+
+    if (!all_unicode) {
+        mp->ma_keys->dk_lookup = lookdict;
+    }
+    mp->ma_keys->dk_usable -= size;
+    mp->ma_used = size;
+    mp->ma_version_tag = DICT_NEXT_VERSION();
+
+    for (Py_ssize_t i = 0; i < size; i++) {
+        MAINTAIN_TRACKING(mp, ep[i].me_key, ep[i].me_value);
```
Comment (Marco-Sulla) [inline Objects/dictobject.c:1710]: Can't be inside the previous loop?

Comment (methane) [reply Objects/dictobject.c:1710]: Note that `PyObject_Hash()` may release GIL.
Once we track the object, GC will scan this object even though its under construction. And this object can be exposed by `gc` module.

I prefer keep it hidden until dict is fully constructed.

File: Objects/dictobject.c (hunk 1 above)
Comment (methane) [reply Objects/dictobject.c:1683]: PyUnicode_CheckExact is simple function. No need to use new flag variable to cache it.

**Pull request closed without merging**